[WGSL] Add code generation for textureSample and vecN

tadeuzagallo opened the pull request:
#### 1ab895ba0e34293c52f4600f84a96ba7802fba8d
<pre>
[WGSL] Add code generation for textureSample and vecN
<a href="https://bugs.webkit.org/show_bug.cgi?id=254433">https://bugs.webkit.org/show_bug.cgi?id=254433</a>
rdar://107193052

Reviewed by Myles C. Maxfield.

Add code in MetalFunctionWriter to rewrite `sampleTexture(t, s, ...)` into `t.sample(s, ...)`
and alias `vecN` to `floatN`. Eventually we might want to perform these manipulations in a
separate phase, but for now this should work.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::visitArguments):
(WGSL::Metal::FunctionDefinitionWriter::visit):

Canonical link: <a href="https://commits.webkit.org/262206@main">https://commits.webkit.org/262206@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9828b696f51973d7854860cf883e47ba94374991

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/916 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/941 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/977 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1326 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/809 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/991 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1026 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1018 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/924 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/861 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/858 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1251 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/855 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/853 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/831 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/882 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1885 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/872 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/819 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/866 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/211 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/886 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -448,21 +461,38 @@ void FunctionDefinitionWriter::visit(AST::CallExpression& call)
+                visitArguments(writer, call, 1);
+            } },
+            { "vec4", ALIAS("float4") },
+        };
```
The indentation doesn't seem to be right

djg replied:
Do we have to use lambdas? What's wrong with a plain-old-function?

tadeuzagallo replied:
> The indentation doesn't seem to be right

Any particular reason? The linter is happy, and I don't see anything wrong. The `#define` can't be indented or the style checker complains, and everything else is indented accordingly as far as I can tell.

> Do we have to use lambdas? What's wrong with a plain-old-function?

We certainly don't have to :) but I thought it was nicer to have everything defined in one place vs the function + the dictionary separate. Is there a reason you would prefer plain functions over lambdas?

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -448,21 +461,38 @@ void FunctionDefinitionWriter::visit(AST::CallExpression& call)
+                writer->m_stringBuilder.append(".sample");
+                visitArguments(writer, call, 1);
+            } },
+            { "vec4", ALIAS("float4") },
```
The macro is only used twice :/

What else will we use it for in the future?

I don't think this needs to be a macro. Can we just pass the string into a lambda?

tadeuzagallo replied:
The macro is only used once now, but it would be used in the future for all the types that have a different name in Metal vs WGSL (at least all the vector and matrix types I believe). I don't think we can pass the string to a lambda, or at least I couldn't find a way to do it, since the lambda has to be constexpr and can't capture anything.

djg commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -91,8 +91,9 @@ class FunctionDefinitionWriter : public AST::Visitor {
     void visitArgumentBufferParameter(AST::Parameter&);
 
-private:
     StringBuilder& m_stringBuilder;
```
Unless it's against the WebKit coding standard, having public members starts with `m_` is a bit on the nose. Can we make it `StringBuilder& stringBuilder`?

tadeuzagallo replied:
I don't think we should mix `m_` and non-`m_` fields, although it's not written anywhere, but I also don't think we should mix public and private `m_` fields. Usually fields are all public, without the `m_` prefix, or all private with the prefix. I'll just revert this change and add an accessor method.

djg reviewed (approved):
:shipit:

tadeuzagallo committed:
[WGSL] Add code generation for textureSample and vecN
https://bugs.webkit.org/show_bug.cgi?id=254433
rdar://107193052

Reviewed by Myles C. Maxfield.

Add code in MetalFunctionWriter to rewrite `sampleTexture(t, s, ...)` into `t.sample(s, ...)`
and alias `vecN` to `floatN`. Eventually we might want to perform these manipulations in a
separate phase, but for now this should work.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::visitArguments):
(WGSL::Metal::FunctionDefinitionWriter::visit):

Canonical link: https://commits.webkit.org/262206@main
